source/common/buffer/watermark_buffer.cc

```diff
@@ -86,7 +89,15 @@ void WatermarkBuffer::checkLowWatermark() {
   below_low_watermark_();
 }
 
-void WatermarkBuffer::checkHighWatermark() {
+void WatermarkBuffer::checkOverflowAndHighWatermarks() {
+  // Check if overflow watermark is enabled, wasn't previously triggered,
+  // and the buffer size is above the threshold
+  if (above_overflow_watermark_ && overflow_watermark_ != 0 && !above_overflow_watermark_called_ &&
+      OwnedImpl::length() > overflow_watermark_) {
+    above_overflow_watermark_called_ = true;
+    above_overflow_watermark_();
```

Comment:
Preference for running overflow watermark callbacks after high watermark callbacks.

Reply:
Just to clarify, in case the buffer length goes from below HW to above OW (and HW), should both the HW and OW be triggered or only the HW will be triggered and in the next verification the OW will be triggered?

Reply:
both should trigger, HW should trigger first then OW.

This is exactly the case why I think we should always call HW first: otherwise we have extra complexity as OW could trigger before or after HW.  HW always first is easier to reason about.